base_external_dbsource_sqlite/models/base_external_dbsource.py

```diff
@@ -52,6 +52,8 @@ def _execute_sqlalchemy(self, sqlquery, sqlparams, metadata):
                     cur = connection.execute(sqlquery)
                 else:
                     cur = connection.execute(sqlquery, sqlparams)
+                if self.env.context.get("no_return", False):
+                    return rows, cols
                 if metadata:
                     cols = list(cur.keys())
                 rows = [r for r in cur]
```

Comment:
Maybe this can be detected automatically?
```suggestion
                try:
                    if metadata:
                        cols = list(cur.keys())
                    rows = [r for r in cur]
                except sqlalchemy.exc.ResourceClosedError:
                    pass
```
This way also avoids returning early if there are multiple records in `self` (though I don't know why there would be).

Reply:
To support multiple records others changes will be needed to avoid the replacement of rows values. 
I agree that could be better but normally this kind of method is called to manually execute a single query that you could know if will be returning values or not, and records in the self record-set refers to a connection datasource. So I don't see the need to try to recover the result records.
Perhaps it's better to split the API methods into select, update and delete, ...

Reply:
> To support multiple records others changes will be needed to avoid the replacement of rows values.

The clobbered rows aren't a problem if you're doing a query without results, though. So returning early could maybe be a problem in exactly these cases.
But it's still unlikely so I guess it doesn't matter much.

> So I don't see the need to try to recover the result records.

The improvement I had in mind is that you don't need to remember to add `no_return` to the context, so it's easier to use.

Reply:
I think that I will close this PR because there are so many design flaws here in this modules. The fact that this method is located on the sqlite specific module is a major flaw.
It will be safer to directly execute:
```python
        with record.connection_open() as connection:
                if sqlparams is None:
                    cur = connection.execute(sqlquery)
                else:
                    cur = connection.execute(sqlquery, sqlparams)
```
without using this method.